src/traces/image/plot.js

```diff
@@ -0,0 +1,146 @@
+/**
+* Copyright 2012-2019, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+var d3 = require('d3');
+var Lib = require('../../lib');
+var xmlnsNamespaces = require('../../constants/xmlns_namespaces');
+var constants = require('./constants');
+
+module.exports = {};
+
+// Generate a function to scale color components according to zmin/zmax and the colormodel
+module.exports.scaler = function(trace) {
+    var colormodel = trace.colormodel;
+    var n = colormodel.length;
+    var cr = constants.colormodel[colormodel];
+
+    function scale(zero, factor, min, max) {
+        return function(c) {
+            c = (c - zero) * factor;
+            c = Lib.constrain(c, min, max);
+            return c;
+        };
+    }
+
+    function constrain(min, max) {
+        return function(c) { return Lib.constrain(c, min, max);};
+    }
+
+    var s = [];
+    // Loop over all color components
+    for(var k = 0; k < n; k++) {
+        if(cr.min[k] !== trace.zmin[k] || cr.max[k] !== trace.zmax[k]) {
+            s.push(scale(
+                trace.zmin[k],
+                (cr.max[k] - cr.min[k]) / (trace.zmax[k] - trace.zmin[k]),
+                cr.min[k],
+                cr.max[k]
+            ));
+        } else {
+            s.push(constrain(cr.min[k], cr.max[k]));
+        }
+    }
+
+    return function(pixel) {
+        var c = pixel.slice();
+        for(var k = 0; k < n; k++) {
+            c[k] = s[k](c[k]);
+        }
+        return c;
+    };
+};
+module.exports.plot = function(gd, plotinfo, cdimage, imageLayer) {
+    var xa = plotinfo.xaxis;
+    var ya = plotinfo.yaxis;
+
+    Lib.makeTraceGroups(imageLayer, cdimage, 'im').each(function(cd) {
```

Comment:
We'll need to fill in 

https://github.com/plotly/plotly.js/blob/6e9bafb8285f7247407a503e7dc33e44ff637f31/src/plots/cartesian/constants.js#L65-L74

with `'imagelayer'`. I'd say image traces should go below `heatmap` traces OR one position above. What do you think @antoinerg 

Moreover, we should add a mock of an image and heatmap trace on the same subplot.

Reply:
> with `'imagelayer'`. I'd say image traces should go below `heatmap` traces OR one position above. What do you think @antoinerg

I don't have a strong preference but like you, I think I would lean to having `image` underneath. This is what I've done in 10c9914c5886de4eb7a9f32a80b747d389c6ce4e

Reply:
Very nicely done @antoinerg !

Thanks!!